src/neural/cuda/inputs_outputs.h

```diff
@@ -65,13 +66,38 @@ struct InputsOutputs {
       ReportCUDAErrors(cudaHostGetDevicePointer(&op_moves_left_mem_gpu_,
                                                 op_moves_left_mem_, 0));
     }
+
+    // memory for network execution managed inside this structure
+    if (tensor_mem_size) {
+      multi_stream_ = true;
+      ReportCUDAErrors(cudaStreamCreate(&stream_));
+      ReportCUDAErrors(cudaMalloc(&scratch_mem_, scratch_size));
+      for (auto& mem : tensor_mem_) {
+        ReportCUDAErrors(cudaMalloc(&mem, tensor_mem_size));
+        ReportCUDAErrors(cudaMemsetAsync(mem, 0, tensor_mem_size, stream_));
+      }
+      ReportCUBLASErrors(cublasCreate(&cublas_));
+      ReportCUBLASErrors(cublasSetMathMode(cublas_, CUBLAS_TENSOR_OP_MATH));
+      ReportCUBLASErrors(cublasSetStream(cublas_, stream_));
```

Comment:
Does this assume that InputOutputs is constructed on the same thread that evaluation is called?  I didn't think that was guaranteed with some backend combinations?

Reply:
Similar but different question.  cudaSetDevice doesn't appear to necessarily have been called for the thread that constructs InputsOutputs - is there potential that this change breaks multi-gpu?

Reply:
if cudaSetDevice isn't called on the thread that constructs InputsOutputs we already would have a problem with multi-gpu as we were allocating some memory there before this change too.

Reply:
I was getting lost in the diff - I've found where cudaSetDevice is called correctly now - so I'm happy now :)